src/MOIWrapper.jl

```diff
@@ -1,408 +1,435 @@
-    (LQOI.SinVar, LQOI.GE),
-    (LQOI.SinVar, LQOI.IV),
-    (LQOI.VecVar, MOI.Nonnegatives),
-    (LQOI.VecVar, MOI.Nonpositives),
-    (LQOI.VecVar, MOI.Zeros),
-    (LQOI.VecLin, MOI.Nonnegatives),
-    (LQOI.VecLin, MOI.Nonpositives),
-    (LQOI.VecLin, MOI.Zeros)
-]
-
-mutable struct Optimizer <: LQOI.LinQuadOptimizer
-    LQOI.@LinQuadOptimizerBase
-    params::Dict{Symbol,Any}
-    Optimizer(::Nothing) = new()
-end
+# Helper functions
+_bounds(s::MOI.GreaterThan{Float64}) = (s.lower, Inf)
+_bounds(s::MOI.LessThan{Float64}) = (-Inf, s.upper)
+_bounds(s::MOI.EqualTo{Float64}) = (s.value, s.value)
+_bounds(s::MOI.Interval{Float64}) = (s.lower, s.upper)
+
+# Supported scalar sets
+const SCALAR_SETS = Union{
+    MOI.GreaterThan{Float64},
+    MOI.LessThan{Float64},
+    MOI.EqualTo{Float64},
+    MOI.Interval{Float64}
+}
+
+mutable struct Optimizer <: MOI.AbstractOptimizer
+    inner::Clp.ClpModel
+    silent::Bool
+
+    optimize_called::Bool
+
+    # TODO: Constructor
+    function Optimizer(;kwargs...)
+        inner_model = Clp.ClpModel()
+        model = new(inner_model, false, false)
+
+        for (key, value) in kwargs
+            MOI.set(model, MOI.RawParameter(key), value)
+        end
 
-### Options
-
-# map option name to C function
-const optionmap = Dict(
-   :PrimalTolerance => set_primal_tolerance,
-   :DualTolerance => set_dual_tolerance,
-   :DualObjectiveLimit => set_dual_objective_limit,
-   :MaximumIterations => set_maximum_iterations,
-   :MaximumSeconds => set_maximum_seconds,
-   :LogLevel => set_log_level,
-   :Scaling => scaling,
-   :Perturbation => set_perturbation,
-   )
-# These options are set by using the ClpSolve object
-const solveoptionmap = Dict(
-   :PresolveType => set_presolve_type,
-   :SolveType => set_solve_type,
-   :InfeasibleReturn => set_infeasible_return,
-   )
-
-function Optimizer(;kwargs...)
-    optimizer = Optimizer(nothing)
-    optimizer.params = Dict{String,Any}()
-    MOI.empty!(optimizer)
-    for (name,value) in kwargs
-        optimizer.params[Symbol(name)] = value
+        return model
     end
-    return optimizer
 end
 
-LQOI.LinearQuadraticModel(::Type{Optimizer},env) = ClpModel()
-
-LQOI.supported_constraints(optimizer::Optimizer) = SUPPORTED_CONSTRAINTS
-LQOI.supported_objectives(optimizer::Optimizer) = SUPPORTED_OBJECTIVES
-
-"""
-    replace_inf(x::Real)
-
-Return `Inf` if `x>1e20`, `-Inf` if `x<-1e20`, and `x` otherwise.
-"""
-function replace_inf(x::Real)
-    if x > 1e20
-        return Inf
-    elseif x < -1e20
-        return -Inf
-    else
-        return x
-    end
+# ====================
+#   empty functions
+# ====================
+function MOI.is_empty(model::Optimizer)
+    # A problem is empty if it has no variable and no linear constraint
+    return (Clp.get_num_rows(model.inner) == 0) && (Clp.get_num_cols(model.inner) == 0)
 end
 
+function MOI.empty!(model::Optimizer)
+    # TODO: keep track of parameters
+    model.inner = Clp.ClpModel()
```

Comment:
Shouldn't we free the previous model there (unless we would like to fully rely on the garbage collector)?

Reply:
Does the function exist in the julia wrapper? Or do we need a direct `ccall`?

Reply:
The following function should do the job:
https://github.com/JuliaOpt/Clp.jl/blob/master/src/ClpCInterface.jl#L230-L237